pkg/reporting/exporters/mongo/mongo.go

```diff
@@ -0,0 +1,147 @@
+package mongo
+
+import (
+	"context"
+	"github.com/pkg/errors"
+	"github.com/projectdiscovery/gologger"
+	"github.com/projectdiscovery/nuclei/v3/pkg/output"
+	"go.mongodb.org/mongo-driver/mongo"
+	"os"
+	"strings"
+	"sync"
+
+	mongooptions "go.mongodb.org/mongo-driver/mongo/options"
+)
+
+type Exporter struct {
+	options    *Options
+	mutex      *sync.Mutex
+	rows       []output.ResultEvent
+	collection *mongo.Collection
+	connection *mongo.Client
+}
+
+// Options contains the configuration options for MongoDB exporter client
+type Options struct {
+	// ConnectionString is the connection string to the MongoDB database
+	ConnectionString string `yaml:"connection-string"`
+	// CollectionName is the name of the MongoDB collection in which to store the results
+	CollectionName string `yaml:"collection-name"`
+	// OmitRaw excludes the Request and Response from the results (helps with filesize)
+	OmitRaw bool `yaml:"omit-raw"`
+	// BatchSize determines the number of results to be kept in memory before writing it to the database or 0 to
+	//	persist all in memory and write all results at the end (default)
+	BatchSize int `yaml:"batch-size"`
+}
+
+// New creates a new MongoDB exporter integration client based on options.
+func New(options *Options) (*Exporter, error) {
+	exporter := &Exporter{
+		mutex:   &sync.Mutex{},
+		options: options,
+		rows:    []output.ResultEvent{},
+	}
+
+	// If the environment variable for the connection string is set, then use that instead. This allows for easier
+	// management of sensitive items such as credentials
+	envConnectionString := os.Getenv("MONGO_CONNECTION_STRING")
+	if envConnectionString != "" {
+		options.ConnectionString = envConnectionString
+		gologger.Info().Msgf("Using connection string from environment variable MONGO_CONNECTION_STRING")
+	}
+
+	// Create the connection to the database
+	clientOptions := mongooptions.Client().ApplyURI(options.ConnectionString)
+
+	// Create a new client and connect to the MongoDB server
+	client, err := mongo.Connect(context.TODO(), clientOptions)
+	if err != nil {
+		gologger.Error().Msgf("Error creating MongoDB client: %s", err)
+	}
+
+	// Ensure the connection is valid
+	err = client.Ping(context.Background(), nil)
+	if err != nil {
+		gologger.Error().Msgf("Error connecting to MongoDB: %s", err)
+	}
+
+	// Get the database from the connection string to set the database and collection by getting the string after the
+	// last / and before the ? options
+	databaseName := strings.Split(options.ConnectionString[strings.LastIndex(options.ConnectionString, "/")+1:], "?")[0]
```

Comment:
I couldn't find a specific parser for MongoDB DSN/conn string, not even in the official client driver package. Splitting strings to parse a DSN can definitely lead to issues, especially when there's no proper test coverage in place. This becomes even more problematic since MongoDB supports two different formats[^1].

Instead of manually splitting the string like this, it's much safer _(& cleaner)_ to use [`url.Parse`](https://pkg.go.dev/net/url#Parse). By parsing the URL, we can directly access the `Path` field and validate it as the DB name.

This way, we're not relying on string manipulation _(which is prone to errors)_, but instead on a well-tested function designed for URL parsing - tho it also future-proofs the parsing a bit in case the conn string format evolves.

[^1]: https://www.mongodb.com/docs/manual/reference/connection-string/#connection-string-formats

Reply:
Switched logic in https://github.com/projectdiscovery/nuclei/pull/5688/commits/fd0045822a314ea22596a2c00bd7cc5de71d3122